packages/whenable/src/whenable.js

```diff
@@ -0,0 +1,111 @@
+// @ts-check
+import { makePromiseKit } from '@endo/promise-kit';
+import { M } from '@endo/patterns';
+import { makeTagged } from '@endo/pass-style';
+
+/**
+ * @param {import('@agoric/base-zone').Zone} zone
+ */
+export const prepareWhenableKits = zone => {
+  /** WeakMap<object, any> */
+  const whenableV0ToEphemeral = new WeakMap();
+
+  /**
+   * Get the current incarnation's promise kit associated with a whenableV0.
+   *
+   * @param {import('./types.js').WhenablePayload['whenableV0']} whenableV0
+   * @returns {import('@endo/promise-kit').PromiseKit<any>}
+   */
+  const findCurrentKit = whenableV0 => {
+    let pk = whenableV0ToEphemeral.get(whenableV0);
+    if (pk) {
+      return pk;
+    }
+
+    pk = makePromiseKit();
+    pk.promise.catch(() => {}); // silence unhandled rejection
+    whenableV0ToEphemeral.set(whenableV0, pk);
+    return pk;
+  };
+
+  /**
+   * @param {'resolve' | 'reject'} kind
+   * @param {import('./types.js').WhenablePayload['whenableV0']} whenableV0
+   * @param {unknown} value
+   */
+  const settle = (kind, whenableV0, value) => {
+    const kit = findCurrentKit(whenableV0);
+    const cb = kit[kind];
+    if (!cb) {
+      return;
+    }
+    whenableV0ToEphemeral.set(whenableV0, harden({ promise: kit.promise }));
+    cb(value);
+  };
+
+  const makeWhenableInternalsKit = zone.exoClassKit(
+    'WhenableInternalsKit',
+    {
+      whenableV0: M.interface('WhenableV0', {
+        shorten: M.call().returns(M.promise()),
+      }),
+      settler: M.interface('Settler', {
+        resolve: M.call().optional(M.any()).returns(),
+        reject: M.call().optional(M.any()).returns(),
+      }),
+    },
+    () => ({}),
+    {
+      whenableV0: {
+        /**
+         * @returns {Promise<any>}
+         */
+        shorten() {
+          return findCurrentKit(this.facets.whenableV0).promise;
+        },
+      },
+      settler: {
```

Comment:
Nit: The use of "settler" throughout this PR isn't really accurate... "settlement" refers to fulfillment or rejection, but is a strict subset of the more general "resolution" that also includes binding a promise to the eventual results of another promise as in e.g. `resolve(E(remotable).method())` (cf. [ECMA-262: Promise Objects](https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-promise-objects)).
```suggestion
      resolvers: {
```

Reply:
Good point, and not a nit. Assuming that the settler object supports forwarding (which I'm reasonably confident it does, by design), then it is not a settler but a resolver. Unfortunately, the noun "resolver" already means a `resolve` function, not an object with `resolve` and `reject` methods. I don't know how to fix this.

If the `resolve` method of a settler does not support forwarding, then I reluctantly suggest it be renamed `fulfill`. But I'm pretty sure it does.

Reply:
Note that https://github.com/tc39/proposal-promise-with-resolvers is now part of standard JavaScript. That suggests that the right noun is the plural `Settlers`. This makes some sense, but this solution would be worse than the problem. When you have a collection of these, you'd want to use the plural of whatever this noun is to name that.

Reply:
If we rename "Whenable" to "Vow", we could rename "Settler" to "VowResolver". One of the advantages of having a very short base name ;)

Reply:
The `resolve` method of a settler supports forwarding (or at least it will once the multi-incarnation bug is fixed), but unusually; `shorten()` leaves it to the consumer to detect if its return value is a promise for another whenable, and if so, to continue traversal from that point.

Reply:
Good. Despite the careful note about `shorten()` (which I appreciate), I think it is accurate enough then to say that `resolve` supports forwarding. And that therefore the purpose of the so-called Settler is indeed resolution, not settlement. What do you think of "VowResolver"?